### PR TITLE
Export supports Training

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94158

Learnings from this prototype
- 

**Joint Graph or Split FW/BW Graphs?** 
- For joint graph to work, everything in the forward pass needs to passed to export(), including the loss function. 
- For a multi-task model, it's unlikely that we export the entire model in one shot. Instead, it's more likely that we export one Module(Arch) at a time. Consequently, the execution flow would be FW_Graph1 ->FW_Graph2 -> BW_Graph2 -> BW_Graph1. It's more feasible to export FW and BW in two separated graphs. 

**Should parameters/buffers be lifted as graph inputs?** 
- AOTAutograd currently lifts all parameters (TODO: not sure about buffers) as graph inputs. It's unclear that if this is right handling for export path: having all the parameters in graph input is confusing for user, and it's result in a different input signature from the original functions.

**Naming UX**
- It's hard to tell original semantics from these names `primals_1, primals_2, tangents_1...`. 
- We need some sort of traceback to link `primals_1` to `self.linear.weight` and `primals_3` to `val`. This is available in node.meta, and it should also be preserved in the serialized graph. 
- Ideally, `primals_1` should have a more meaningful name, e.g. `self_linear_weight`.  `tangents_1` should be `loss_grad`. 

**Implementation Notes**
- ATM, inference export is calling make_fx, this training export prototype is calling aot_module_simplified. We should unify them, so that they have the identical graph properties, e.g. handling of parameters. 
- There are merits for torch.compile() and torch.export() to share the aot_module_simplified path, but if their technical requirements diverges, we should consider adding an export-specific path for AOTAutograd. 

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/eval_frame.py [hunk 1]
```diff
@@ -646,7 +646,45 @@ def run_node(self, n):
...
+            fw_compiler=graph_saver_forward if training_mode == "split" else None,
+            bw_compiler=graph_saver_backward if training_mode == "split" else None,
+            partition_fn=graph_saver_joint if training_mode == "joint" else None,
+            decompositions=decomposition_table,
+        )
+
+        if training_mode == "joint":
+            return (joint_gm, out_guards)
+        elif training_mode == "split":
+            return ((fw_gm, bw_gm), out_guards)
```
Comment (SherlockNoMad) [inline torch/_dynamo/eval_frame.py:685]: Need better return UX.

File: torch/_dynamo/eval_frame.py [hunk 2]
```diff
@@ -646,7 +646,45 @@ def run_node(self, n):
...
+
+    elif aten_graph:
```
Comment (SherlockNoMad) [inline torch/_dynamo/eval_frame.py:687]: Should get rid of this path.

File: test/dynamo/test_export.py [hunk 3]
```diff
@@ -1752,6 +1752,94 @@ def fn_with_kwargs(pos0, tuple0, *myargs):
...
+                loss = out.sum()
+                return loss
+
+        mod = Module()
+        x = torch.randn([3, 3], requires_grad=True)
+
+        torch._dynamo.reset()
+        joint_gm, _  = torch._dynamo.export(mod, x, aten_graph=True, training_mode='joint', tracing_mode='symbolic')
+
+        self.assertExpectedInline(joint_gm.print_readable(), """\
```
Comment (jansel) [inline test/dynamo/test_export.py:1773]: Doing tests this way will be brittle as if we change our printing logic or variable naming logic or some small part of the trace they will fail.

Could we instead test the functionality of the graph?

**Review by jansel — CHANGES_REQUESTED** [review]

File: test/dynamo/test_export.py (hunk 3 above)
Comment (ezyang) [reply test/dynamo/test_export.py:1773]: So, the point of expect tests is that they're easy to update (you just run with `EXPECTTEST_ACCEPT=1` and the source files are all updated with the new values). That being said, we can't land *these* tests, they have absolute filepaths hardcoded in them; they'll have to be normalized at least.

Comment (SherlockNoMad) [reply test/dynamo/test_export.py:1773]: I don't plan to check this test in. Just showing the existing exported graph to discuss what are the changes we should make. In particular, we need to design the properties, invariants, conventions of exported training graphs.
